### PR TITLE
Bursts

### PR DESCRIPTION
**Pull request by thattil**

### CODE REVIEW

File: slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c [hunk 1]
```diff
@@ -748,12 +758,39 @@ int64_t getNumFrames() {
 void setNumTriggers(int64_t val) {
     if (val > 0) {
 		FILE_LOG(logINFO, ("Setting number of triggers %lld\n", (long long int)val));
```
Comment (erikfrojdh) [inline slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c]: I would say that the cast is not needed

File: slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c [hunk 2]
```diff
@@ -748,12 +758,39 @@ int64_t getNumFrames() {
...
+}
+
+int64_t getNumBursts() {
+	if (burstMode != BURST_OFF && getTiming() == AUTO_TIMING) {
+		return get64BitReg(SET_CYCLES_LSB_REG, SET_CYCLES_MSB_REG);
+	}
+	return numBursts;
 }
 
 int setExpTime(int64_t val) {
```
Comment (erikfrojdh) [inline slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c]: Not displayed here but looks like setExptimeBurst and setExptimeCont calls the same function. Maybe I misunderstood the logic but seems confusing.

File: slsDetectorSoftware/src/slsDetector.cpp [hunk 3]
```diff
@@ -1253,6 +1279,29 @@ void slsDetector::setNumberOfTriggers(int64_t value) {
     shm()->nTriggers = value;
     sendTotalNumFramestoReceiver();
 }
+
+int64_t slsDetector::getNumberOfBurstsFromShm() {
```
Comment (erikfrojdh) [inline slsDetectorSoftware/src/slsDetector.cpp]: Is this function used? When searching I don't find it...

File: slsDetectorSoftware/src/slsDetector.cpp [hunk 4]
```diff
@@ -1253,6 +1279,29 @@ void slsDetector::setNumberOfTriggers(int64_t value) {
...
+    return shm()->nBursts; 
+}
+
+int64_t slsDetector::getNumberOfBursts() {
+    int64_t prevVal = shm()->nBursts;
+    int64_t retval = -1;
+    sendToDetector(F_GET_NUM_BURSTS, nullptr, retval);
+    FILE_LOG(logDEBUG1) << "number of bursts :" << retval;
+    shm()->nBursts = retval;
+    if (prevVal != retval) {
```
Comment (erikfrojdh) [inline slsDetectorSoftware/src/slsDetector.cpp]: by comparing against shm()->nBursts you can skip using prevVal

File: slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c (hunk 2 above)
Comment (thattil) [reply slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c]: Yes. This is for now. Soon, the plan is to have different registers.

File: slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c (hunk 1 above)
Comment (thattil) [reply slsDetectorServers/gotthard2DetectorServer/slsDetectorFunctionList.c]: done

File: slsDetectorSoftware/src/slsDetector.cpp (hunk 3 above)
Comment (thattil) [reply slsDetectorSoftware/src/slsDetector.cpp]: I missed the progress by multi. Removed all the getNumberofFramesfromShm, triggersfromshm, and burstsfromshm.. and made into one function to get getTotalNumFramestoReceive.

File: slsDetectorSoftware/src/slsDetector.cpp (hunk 4 above)
Comment (thattil) [reply slsDetectorSoftware/src/slsDetector.cpp]: but i need to update shm()->nBursts before calling sendTotalNumFrametoReceiver().
If I do it before the if condition, I need a new variable.
If I do it inside the if condition, I need to repeat the asignment in the else.

Comment (thattil) [reply slsDetectorSoftware/src/slsDetector.cpp]: done

**Comment by thattil** [comment]
The nbursts is needed to calculate progress in multi. It is used instead of triggers in gotthard2 for auto and burst mode.

**Pull request merged**